sdk/core/core-https/review/core-https.api.md

```diff
@@ -4,8 +4,53 @@
 
 ```ts
 
-// @public (undocumented)
-export function helloWorld(): string;
+// @public
+export interface AddPipelineOptions {
+    afterPhase?: PipelinePhase;
+    afterPolicies?: string[];
+    beforePolicies?: string[];
+    phase?: PipelinePhase;
+}
+
+// @public
+export interface HttpsClient {
+    sendRequest: SendRequest;
+}
+
+// @public
+export interface Pipeline {
+    addPolicy(policy: PipelinePolicy, options?: AddPipelineOptions): void;
+    clone(): Pipeline;
+    getOrderedPolicies(): PipelinePolicy[];
+    removePolicy(options: {
+        name?: string;
+        phase?: PipelinePhase;
+    }): PipelinePolicy[];
+    sendRequest(httpsClient: HttpsClient, request: PipelineRequest): Promise<PipelineResponse>;
+}
+
+// @public
+export type PipelinePhase = "Serialize" | "Retry";
```

Comment:
I recently ran into an issue when needing to add a custom policy to storage-blob clients. The problem was I needed to:

1. set a header after serialization (because this header was set with a different value than I wanted during serialization)
2. needed to make sure it happened before a signing policy since the signer took the header value into account

It looks like there are a lot of different ways to accomplish this given what you have here, and I'm not sure what the right way is.

Would it be possible to compile a list of our existing policies and how they'd fit into these phases? I don't think that has to be done in this PR (maybe it just goes into the issue) since we can always add/change the phases.

Reply:
> It looks like there are a lot of different ways to accomplish this given what you have here, and I'm not sure what the right way is.

Naively, it sounds like you'd want ` { afterPhase: 'Serialize', beforePolicies: [SigningPolicy.name] }`

It's also fine if you want to name the serialization policies explicitly with `afterPolicies`, but phases are intended to help you run after well-known groups (packing/unpacking messages in the case of Serialize and re-issuing failed but recoverable requests in case of Retry)

> 
> Would it be possible to compile a list of our existing policies and how they'd fit into these phases? I don't think that has to be done in this PR (maybe it just goes into the issue) since we can always add/change the phases.

Glancing at `createDefaultRequestPolicyFactories` in `ServiceClient` I can imagine something like:

 * Unassigned:
   * Client request ID
   * Auth
   * User agent
   * Redirect
* Serialize:
  * deserialization (json/xml)
  * mapping (codegen assisted)
* Retry:
  * Exponential
  * System error
  * Throttling
* afterPhase: retry
  * proxy
  * logging

Does that help?